Framework/API/src/ExperimentInfo.cpp

```diff
@@ -176,6 +184,24 @@ void checkDetectorInfoSize(const Instrument &instr,
                              "DetectorInfo and number of detectors in "
                              "instrument");
 }
+}
+
+boost::shared_ptr<Beamline::ComponentInfo>
+makeComponentInfo(const Instrument &oldInstr,
+                  const API::DetectorInfo &detectorInfo,
+                  std::vector<Geometry::ComponentID> &componentIds) {
+
+  InfoComponentVisitor visitor(detectorInfo);
+  std::vector<size_t> detectorIndexes; // Not strictly needed at this level
+
+  // Register everything via visitor
+  oldInstr.registerContents(visitor, detectorIndexes);
+  // Extract component ids. We need this for the ComponentInfo wrapper.
+  componentIds = visitor.componentIds();
+
+  return boost::make_shared<Mantid::Beamline::ComponentInfo>(
+      visitor.componentDetectorIndexes());
```

Comment:
Minor point, but by having `componentIds` and `componentDetectorIndices` as members of the visitor we are doing an extra copy here.